### PR TITLE
FHIR Mortality masking provider

### PR DESCRIPTION
**Pull request by jarrettranderson**
Add new FHIR Mortality masking provider.
Correct issue with identification of nodes in JSON trees when node objects are reused to represent the same value in multiple places in the tree.

### CODE REVIEW

File: docs/masking-providers.md [hunk 1]
```diff
@@ -733,6 +734,36 @@ date at */dateWritten*.
...
 
+#### FHIR_MORTALITY_DEPENDENCY
+
+This provider enforces a specific type of dependency between the birth date and the deceased datetime and deceased boolean properties that exist in a FHIR Patient resource, which is a very specific schema imposed on a JSON document. 
+
+This provider may only be applied to properties named either exactly "deceasedDateTime" or "deceasedBoolean".  When using this provider, it is recommended to configure the provider on both of these properties. It is expected that both of these properties and another property named "birthDate" all belong to the same parent object in the JSON document. 
+
+When the privacy provider is applied to "deceasedDateTime" and "deceasedDateTime" has a non-null value in the input document, the value of "deceasedDateTime" is set to null.  In addition, the value of "deceasedBoolean" is either set to _true_ or _null_ depending upon the age of this patient when death occurred.  If "deceasedBoolean" does not exist as a sibling of "deceasedDateTime", it will be added to the JSON structure.
```
Comment (dumashwang) [inline docs/masking-providers.md:743]: "depending upon the age of this patient when death occurred". 

At what age is deceasedBoolean set to _true_?

File: docs/masking-providers.md [hunk 2]
```diff
@@ -733,6 +734,36 @@ date at */dateWritten*.
 | preserveDomains | Integer  | Number of domains to preserve starting from the right; if value = -1 then all domains will be preserved (that is, everything after \@) | 1                 |
 | nameLength      | Integer  | Length of username to generate; if value = -1, then the username will be 5 to 8 characters long. 0 is invalid      | \-1               |
 
+#### FHIR_MORTALITY_DEPENDENCY
+
+This provider enforces a specific type of dependency between the birth date and the deceased datetime and deceased boolean properties that exist in a FHIR Patient resource, which is a very specific schema imposed on a JSON document.
```
Comment (dumashwang) [inline docs/masking-providers.md:739]: "This provider enforces a specific type of dependency between the birth date and the deceased datetime and deceased boolean properties"

Is this provider enforcing a specification in FHIR?  If so, can you provide a link to the part of the specification this is trying to enforce?

File: ipv-core/src/main/java/com/ibm/whc/deid/providers/masking/AbstractMaskingProvider.java [hunk 3]
```diff
@@ -110,22 +112,39 @@ public void maskIdentifierBatch(List<MaskingActionInputIdentifier> identifiers)
...
+      } catch (NumberFormatException e) {
+        // should not occur
+        throw new RuntimeException("offset in path cannot be converted to an integer");
+      }
+      try {
+        ((ArrayNode) i.getParent()).set(offset, newNode);
+      } catch (IndexOutOfBoundsException e) {
+        // should not occur
+        throw new RuntimeException("offset in path was " + offset + " but ArrayNode has only "
+            + ((ArrayNode) i.getParent()).size() + " members");
```
Comment (dumashwang) [inline ipv-core/src/main/java/com/ibm/whc/deid/providers/masking/AbstractMaskingProvider.java:146]: Seems to have many branches(if/try), and I don't see unit tests added for AbstractMaskingProvider.  We can consider adding unit tests for the future.

File: ipv-core/src/main/java/com/ibm/whc/deid/providers/masking/fhir/FHIRMortalityDependencyMaskingProvider.java [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+          }
+        } else {
+          month = 1;
+        }
+        date = LocalDate.of(year, month, 1);
+      } else {
+        // there is enough data provided for LocaleDate to parse the value
+        date = LocalDate.parse(stringValue);
+      }
+    } catch (Exception e) {
```
Comment (dumashwang) [inline ipv-core/src/main/java/com/ibm/whc/deid/providers/masking/fhir/FHIRMortalityDependencyMaskingProvider.java:234]: Is it possible to catch a more specific exception here?  Do we need to catch all exceptions?

**Review by dumashwang — APPROVED** [review]
Looks good.  I added some comments.  Please review.

File: docs/masking-providers.md (hunk 1 above)
Comment (jarrettranderson) [reply docs/masking-providers.md:743]: The following paragraphs explain about the age threshold - basically its a configurable number of years.  If you are older than that, you can have deceasedBoolean=true.  If not, you get deceasedBoolean=null.

File: docs/masking-providers.md (hunk 2 above)
Comment (jarrettranderson) [reply docs/masking-providers.md:739]: This is not a specifc FHIR specification.  It is based on looking at the current FHIR 4 documentation about what the fields are and what their content is.  The interaction between the fields is based on the customer request.

File: ipv-core/src/main/java/com/ibm/whc/deid/providers/masking/AbstractMaskingProvider.java (hunk 3 above)
Comment (jarrettranderson) [reply ipv-core/src/main/java/com/ibm/whc/deid/providers/masking/AbstractMaskingProvider.java:146]: Most of these checks are not actually necessary.  We create the strings its examining internally in the code, so we would expect them to be formatted correctly.  Its not direct data from the user.  All of our existing unit tests that to end-to-end masking flow through this method to update the JSON document.

File: ipv-core/src/main/java/com/ibm/whc/deid/providers/masking/fhir/FHIRMortalityDependencyMaskingProvider.java (hunk 4 above)
Comment (jarrettranderson) [reply ipv-core/src/main/java/com/ibm/whc/deid/providers/masking/fhir/FHIRMortalityDependencyMaskingProvider.java:234]: There are a few different types of unrelated exceptions possible here.  I could have listed them all in the catch, but I believe we would want to handle any exception even unforeseen ones in this same way, so Exception seemed appropriate.

Comment (jarrettranderson) [reply ipv-core/src/main/java/com/ibm/whc/deid/providers/masking/fhir/FHIRMortalityDependencyMaskingProvider.java:234]: There are a few different Exception types possible from this block.  I could have listed those in the catch, but I believe we would want to catch any exceptions in this block, even unforeseen ones, in the same way so I thought Exception was appropriate here.

**Pull request merged**